scalagen/src/test/scala/org/scalameta/scalagen/Examples.scala

```diff
@@ -27,10 +27,52 @@ class PrintHi extends ExtensionGenerator("PrintHi") {
   }
 }
 
-class TestRecurse extends ExtensionGenerator("TestRecurse") {
+case class PrintHiInCompanion() extends CompanionGenerator("PrintHiInCompanion") {
+  override def extendCompanion(c: Defn.Class): List[Stat] = {
+    val hi: Lit.String = Lit.String("hi")
+    val hiMethod: Defn.Def =
+      q"def hi = println($hi)"
+
+    hiMethod :: Nil
+  }
+}
+
+case class TestRecurse() extends ExtensionGenerator("TestRecurse") {
```

Comment:
How does it simplify semantic discovery? The FQN is specified by the generator. The semantic discovery must build a semantic database of the current state of the Tree (It changes every transformation, regardless of whether recursion is involved.)

I'm happy to remove recursion if it proves to be an issue. But I do not see it being any more of a problem than the current behaviour.

Reply:
> The semantic discovery must build a semantic database of the current state of the Tree (It changes every transformation, regardless of whether recursion is involved.)

I had in mind that we should avoid this. We can provide a semantic view on the first expansion but not after that. Semantic information for macro annotations is a hard problem https://contributors.scala-lang.org/t/annotation-macros/1211/18 We can side-step the problem by enforcing the limitation that the un-expanded code can't depend on the expanded code. It will be harder to enforce this limitation and provide helpful error messages if we typecheck half-expanded code.

Reply:
Hmmmm.... It might be possible. I have an idea! I'll create a new ticket.